This is the error I am getting when I try to enable the python bindings:

```
--- stderr: blmc_controllers                                 
CMake Warning at /home/jviereck_local/dev/learned_endeff_tracking/workspace/install/pybind11/share/cmake/pybind11/pybind11Tools.cmake:131 (add_library):
  Cannot generate a safe runtime search path for target py_blmc_controllers
  because files in some directories may conflict with libraries in implicit
  directories:

    runtime library [liburdfdom_sensor.so.1.0] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /opt/ros/eloquent/lib
    runtime library [liburdfdom_model_state.so.1.0] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /opt/ros/eloquent/lib
    runtime library [liburdfdom_model.so.1.0] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /opt/ros/eloquent/lib
    runtime library [liburdfdom_world.so.1.0] in /usr/lib/x86_64-linux-gnu may be hidden by files in:
      /opt/ros/eloquent/lib

  Some of these libraries may not be found correctly.
Call Stack (most recent call first):
  CMakeLists.txt:71 (pybind11_add_module)


/usr/bin/ld: libblmc_controllers.a(impedance_controller.cpp.o): relocation R_X86_64_PC32 against symbol `_ZN9pinocchio8ModelTplIdLi0ENS_25JointCollectionDefaultTplEEC1Ev' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -8,7 +8,7 @@ project(blmc_controllers)
 if(NOT ${CMAKE_SYSTEM_NAME} MATCHES "Darwin")
   set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wl,--no-as-needed")
 endif()
-set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -Wall -W -Wextra")
+set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -Wall -W -Wextra -fPIC")
 
 # ensuring path to libraries are set during install
 set(CMAKE_SKIP_BUILD_RPATH  FALSE)
@@ -37,6 +37,7 @@ find_package(yaml_utils REQUIRED)
 
 add_library(${PROJECT_NAME}
     src/impedance_controller.cpp
+    src/centroidal_pd_controller.cpp
 )
 target_link_libraries(${PROJECT_NAME} pinocchio::pinocchio)
 target_link_libraries(${PROJECT_NAME} Eigen3::Eigen)
@@ -68,17 +69,19 @@ install(DIRECTORY include/ DESTINATION include)
 # ################################
 # # python bindings with pybind11#
 # ################################
-# pybind11_add_module(py_${PROJECT_NAME} MODULE
-#     srcpy/${PROJECT_NAME}.cpp
-#     srcpy/impedance_controller.cpp
-# )
+pybind11_add_module(py_${PROJECT_NAME} MODULE
+    srcpy/${PROJECT_NAME}.cpp
+    srcpy/impedance_controller.cpp
+    srcpy/centroidal_pd_controller.cpp
+)
 
-# target_link_libraries(py_${PROJECT_NAME} PRIVATE pybind11::module)
-# target_link_libraries(py_${PROJECT_NAME} PRIVATE ${PROJECT_NAME})
+
+target_link_libraries(py_${PROJECT_NAME} PRIVATE pybind11::module)
+target_link_libraries(py_${PROJECT_NAME} PRIVATE ${PROJECT_NAME})
 
 # # install the bindings
 get_python_install_dir(PYTHON_INSTALL_DIR)
-# install(TARGETS py_${PROJECT_NAME} DESTINATION ${PYTHON_INSTALL_DIR})
+install(TARGETS py_${PROJECT_NAME} DESTINATION ${PYTHON_INSTALL_DIR})
 
 ################################
 # install python packages      #

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -26,6 +26,7 @@ set(CMAKE_INSTALL_RPATH ${CMAKE_INSTALL_PREFIX}/lib)
 find_package(pybind11 CONFIG REQUIRED)
 find_package(Eigen3 REQUIRED)
 find_package(pinocchio REQUIRED)
+find_package(eiquadprog REQUIRED)
 
 #local depends
 find_package(mpi_cmake_modules REQUIRED)
@@ -38,9 +39,11 @@ find_package(yaml_utils REQUIRED)
 add_library(${PROJECT_NAME}
     src/impedance_controller.cpp
     src/centroidal_pd_controller.cpp
+    src/centroidal_force_qp_controller.cpp
 )
 target_link_libraries(${PROJECT_NAME} pinocchio::pinocchio)
 target_link_libraries(${PROJECT_NAME} Eigen3::Eigen)
+target_link_libraries(${PROJECT_NAME} eiquadprog::eiquadprog)
 target_link_libraries(${PROJECT_NAME} yaml_utils::yaml_utils)
 
 # Includes. Add the include dependencies
@@ -73,6 +76,7 @@ pybind11_add_module(py_${PROJECT_NAME} MODULE
     srcpy/${PROJECT_NAME}.cpp
     srcpy/impedance_controller.cpp
     srcpy/centroidal_pd_controller.cpp
+    srcpy/centroidal_force_qp_controller.cpp
 )
 
 

--- a/Config.cmake.in
+++ b/Config.cmake.in
@@ -1,0 +1,11 @@
+@PACKAGE_INIT@
+
+include("${CMAKE_CURRENT_LIST_DIR}/blmc_controllersTargets.cmake")
+
+include(CMakeFindDependencyMacro)
+
+# we do not add the other dependencies because these are header files lib
+find_dependency(yaml-cpp CONFIG REQUIRED)
+find_dependency(Eigen3 REQUIRED)
+
+check_required_components(blmc_controllers)

--- a/include/blmc_controllers/centroidal_force_qp_controller.hpp
+++ b/include/blmc_controllers/centroidal_force_qp_controller.hpp
@@ -1,0 +1,85 @@
+/**
+ * @file
+ * @license BSD 3-clause
+ * @copyright Copyright (c) 2020, New York University and Max Planck
+ * Gesellschaft
+ *
+ * @brief Implements a qp allocating forces to track a desired centroidal wrench.
+ *
+ */
+
+#pragma once
+
+#include <Eigen/Dense>
+#include <eiquadprog/eiquadprog-fast.hpp>
+
+namespace blmc_controllers
+{
+typedef Eigen::Array<double, 6, 1> Array6d;
+typedef Eigen::Matrix<double, 6, 1> Vector6d;
+
+/**
+ * @brief Impedance controller between any two frames of the robot.
+ */
+class CentroidalForceQPController
+{
+public:
+    /**
+     * @brief Construct a new ImpedanceController object.
+     */
+    CentroidalForceQPController();
+
+    /**
+     * @brief Initialize the internal data. None real-time safe method.
+     *
+     * @param number_endeffectors Maximum number of endeffectors in the problem.
+     * @param friction_coeff Floor friction coefficient to use.
+     * @param qp_penalty_lin The penalty weight for the linear wcom violation.
+     * @param qp_penalty_ang The penalty weight for the angular wcom violation.
+     */
+    void initialize(int number_endeffectors, double friction_coeff,
+                    double qp_penalty_lin, double qp_penalty_ang);
+
+    /**
+     * @brief Computes the centroidal wrench using a PD controller.
+     * 
+     * @param w_com The desired centroidal wrench to track.
+     * @param relative_position_endeff The relative position of the endeffectors
+     *     with respect to the center of mass.
+     */
+    void run(Eigen::Ref<const Vector6d> w_com,
+             Eigen::Ref<const Eigen::VectorXd> relative_position_endeff,
+             Eigen::Ref<const Eigen::VectorXd> cnt_array);
+
+    /**
+     * @brief Get the computed desired forces
+     *
+     * @return Eigen::VectorXd&
+     */
+    Eigen::VectorXd& get_forces();
+
+private:  // attributes
+    /** @brief Output forces */
+    Eigen::VectorXd forces_;
+    Eigen::VectorXd sol_;
+
+    Eigen::MatrixXd hess_;
+    Eigen::MatrixXd ce_;
+    Eigen::MatrixXd ce_new_;
+    Eigen::MatrixXd ci_;
+
+    Eigen::VectorXd g0_;
+    Eigen::VectorXd ci0_;
+
+    int nb_eff_;
+
+    double mu_;
+
+    double qp_penalty_lin_;
+    double qp_penalty_ang_;
+
+    eiquadprog::solvers::EiquadprogFast qp_;
+
+};
+
+}  // namespace blmc_controllers

--- a/include/blmc_controllers/centroidal_pd_controller.hpp
+++ b/include/blmc_controllers/centroidal_pd_controller.hpp
@@ -1,0 +1,83 @@
+/**
+ * @file
+ * @license BSD 3-clause
+ * @copyright Copyright (c) 2020, New York University and Max Planck
+ * Gesellschaft
+ *
+ * @brief Implements a PD controller at the center of mass.
+ *
+ */
+
+#pragma once
+
+#include <Eigen/Dense>
+
+namespace blmc_controllers
+{
+typedef Eigen::Array<double, 6, 1> Array6d;
+typedef Eigen::Matrix<double, 6, 1> Vector6d;
+
+/**
+ * @brief Impedance controller between any two frames of the robot.
+ */
+class CentroidalPDController
+{
+public:
+    /**
+     * @brief Construct a new ImpedanceController object.
+     */
+    CentroidalPDController();
+
+    /**
+     * @brief Initialize the internal data. None real-time safe method.
+     *
+     * @param inertia Inertia of the base.
+     */
+    void initialize(double& mass, Eigen::Ref<const Eigen::Vector3d> inertia);
+
+    /**
+     * Computes the centroidal wrench using a PD controller.
+     */
+    void run(Eigen::Ref<const Eigen::Vector3d> kc,
+             Eigen::Ref<const Eigen::Vector3d> dc,
+             Eigen::Ref<const Eigen::Vector3d> kb,
+             Eigen::Ref<const Eigen::Vector3d> db,
+             Eigen::Ref<const Eigen::Vector3d> com,
+             Eigen::Ref<const Eigen::Vector3d> com_des,
+             Eigen::Ref<const Eigen::Vector3d> vcom,
+             Eigen::Ref<const Eigen::Vector3d> vcom_des,
+             Eigen::Ref<const Eigen::Vector4d> ori,
+             Eigen::Ref<const Eigen::Vector4d> ori_des,
+             Eigen::Ref<const Eigen::Vector3d> angvel,
+             Eigen::Ref<const Eigen::Vector3d> angvel_des);
+
+    /**
+     * @brief Get the computed wrench from the PD controller.
+     *
+     * @return Eigen::VectorXd&
+     */
+    Vector6d& get_wrench();
+
+private:  // attributes
+    /** @brief Output wrench */
+    Vector6d wrench_;
+
+    double mass_;
+    Eigen::Vector3d inertia_;
+
+    Eigen::Vector3d pos_error_;
+    Eigen::Vector3d vel_error_;
+    Eigen::Vector3d ori_error_;
+
+    Eigen::Quaternion<double> ori_quat_;
+    Eigen::Quaternion<double> des_ori_quat_;
+    Eigen::Quaternion<double> ori_error_quat_;
+
+
+    Eigen::Matrix<double, 3, 3> ori_se3_;
+    Eigen::Matrix<double, 3, 3> des_ori_se3_;
+    Eigen::Matrix<double, 3, 3> ori_error_se3_; // refer to christian ott paper for definitions (Rdb)
+    
+};
+
+}  // namespace blmc_controllers

--- a/python/blmc_controllers/robot_centroidal_controller.py
+++ b/python/blmc_controllers/robot_centroidal_controller.py
@@ -120,7 +120,7 @@ class RobotCentroidalController:
             A[3:, 3 * j:3 * (j + 1)] = pin.skew(r[i])
 
             G[5*j + 0, 3 * j + 0] = 1    # mu Fz - Fx >= 0
-            G[5*j + 0, 3 * j + 2] = -self.robot_mass
+            G[5*j + 0, 3 * j + 2] = -self.mu
             G[5*j + 1, 3 * j + 0] = -1     # mu Fz + Fx >= 0
             G[5*j + 1, 3 * j + 2] = -self.mu
             G[5*j + 2, 3 * j + 1] = 1    # mu Fz - Fy >= 0

--- a/src/centroidal_force_qp_controller.cpp
+++ b/src/centroidal_force_qp_controller.cpp
@@ -1,0 +1,138 @@
+/**
+ * @file
+ * @license BSD 3-clause
+ * @copyright Copyright (c) 2020, New York University and Max Planck
+ * Gesellschaft
+ *
+ * @brief Implements a PD controller at the center of mass.
+ *
+ */
+
+#include <iostream>
+#include "blmc_controllers/centroidal_force_qp_controller.hpp"
+
+namespace blmc_controllers
+{
+
+
+CentroidalForceQPController::CentroidalForceQPController() 
+{
+}
+
+void CentroidalForceQPController::initialize(
+    int number_endeffectors, double friction_coeff,
+    double qp_penalty_lin, double qp_penalty_ang)
+{
+    nb_eff_ = number_endeffectors;
+    mu_ = friction_coeff;
+    qp_penalty_lin_ = qp_penalty_lin;
+    qp_penalty_ang_ = qp_penalty_ang;
+
+    // Resize the problem.
+    qp_.reset(3 * nb_eff_ + 6, 9, 5 * nb_eff_);
+    sol_.resize(3 * nb_eff_ + 6);
+    forces_.resize(3 * nb_eff_);
+    sol_.fill(0.);
+    forces_.fill(0.);
+
+    // The QP solves the following problem.
+    //
+    // min. 0.5 * x' Hess x + g0' x
+    // s.t. CE x + ce0 = 0
+    //      CI x + ci0 >= 0
+
+    hess_.resize(3 * nb_eff_ + 6, 3 * nb_eff_ + 6);
+    g0_.resize(3 * nb_eff_ + 6);
+
+    ce_.resize(6, 3 * nb_eff_ + 6);
+    ce_new_.resize(6, 3 * nb_eff_ + 6);
+
+    ci_.resize(5 * nb_eff_, 3 * nb_eff_ + 6);
+    ci0_.resize(5 * nb_eff_);
+
+    // Simple initializations.
+    g0_.fill(0.);
+    ci_.fill(0.);
+    ci0_.fill(0.);
+    ce_.fill(0.);
+    hess_.setIdentity();
+    hess_ *= 2.;
+
+    // Slack variables weights.
+    hess_.block<3, 3>(3 * nb_eff_, 3 * nb_eff_).setIdentity();
+    hess_.block<3, 3>(3 * nb_eff_, 3 * nb_eff_) *= qp_penalty_lin;
+    hess_.block<3, 3>(3 * nb_eff_ + 3, 3 * nb_eff_ + 3).setIdentity();
+    hess_.block<3, 3>(3 * nb_eff_ + 3, 3 * nb_eff_ + 3) *= qp_penalty_ang;
+
+    // Setup centroidal wrench equality.
+    for (int i = 0; i < nb_eff_; i++)
+    {
+        // Setup the linear part. The angular part with the cross product is
+        // setup in the run() method.
+        ce_.block<3, 3>(0, 3 * i) << -1,  0,  0,
+                                      0, -1,  0,
+                                      0,  0, -1;
+    }
+
+    // Part of the slack variables.
+    ce_.block<6, 6>(0, 3 * nb_eff_).setIdentity();
+
+    // Setup the friction cone constraints.
+    for (int j = 0; j < nb_eff_; j++)
+    {
+        ci_(5 * j + 0, 3 * j + 0) = -1;    // mu Fz - Fx >= 0
+        ci_(5 * j + 0, 3 * j + 2) = mu_;
+        ci_(5 * j + 1, 3 * j + 0) = 1;     // mu Fz + Fx >= 0
+        ci_(5 * j + 1, 3 * j + 2) = mu_;
+        ci_(5 * j + 2, 3 * j + 1) = -1;    // mu Fz - Fy >= 0
+        ci_(5 * j + 2, 3 * j + 2) = mu_;
+        ci_(5 * j + 3, 3 * j + 1) = 1;     // mu Fz + Fy >= 0
+        ci_(5 * j + 3, 3 * j + 2) = mu_;
+        ci_(5 * j + 4, 3 * j + 2) = 1;     // Fz >= 0
+    }
+}
+
+void CentroidalForceQPController::run(
+    Eigen::Ref<const Vector6d> w_com,
+    Eigen::Ref<const Eigen::VectorXd> relative_position_endeff,
+    Eigen::Ref<const Eigen::VectorXd> cnt_array
+) 
+{
+    // Copy the linear part for the centroidal wrench equality.
+    ce_new_ = ce_;
+
+    // Setup cross product at the endeffectors.
+    for (int i = 0; i < nb_eff_; i++) 
+    {
+        double x = relative_position_endeff(3 * i);
+        double y = relative_position_endeff(3 * i + 1);
+        double z = 0; // Always assumed to be on the floor.
+        ce_new_.block<3, 3>(3, 3 * i) << 0,  z, -y,
+                                        -z,  0,  x,
+                                         y, -x,  0;
+    }
+
+    // Deactivate forces not in contact with the ground.
+    for (int i = 0; i < nb_eff_; i++)
+    {
+        if (cnt_array(i) < 0.2)
+        {
+            ce_new_.col(3 * i + 0).setZero();
+            ce_new_.col(3 * i + 1).setZero();
+            ce_new_.col(3 * i + 2).setZero();
+        }
+    }
+
+    // Solve the QP.
+    std::cout << "Solution result: ";
+    auto status = qp_.solve_quadprog(hess_, g0_, ce_new_, w_com, ci_, ci0_, sol_);
+    std::cout << status << std::endl;
+    forces_ = sol_.head(3 * nb_eff_);
+}
+
+Eigen::VectorXd& CentroidalForceQPController::get_forces()
+{
+    return forces_;
+}
+
+}  // namespace blmc_controllers

--- a/src/centroidal_pd_controller.cpp
+++ b/src/centroidal_pd_controller.cpp
@@ -1,0 +1,90 @@
+/**
+ * @file
+ * @license BSD 3-clause
+ * @copyright Copyright (c) 2020, New York University and Max Planck
+ * Gesellschaft
+ *
+ * @brief Implements a PD controller at the center of mass.
+ *
+ */
+
+#include "blmc_controllers/centroidal_pd_controller.hpp"
+
+namespace blmc_controllers
+{
+
+
+CentroidalPDController::CentroidalPDController() 
+{
+}
+
+void CentroidalPDController::initialize(double& mass, 
+                                   Eigen::Ref<const Eigen::Vector3d> inertia)
+{
+    mass_ = mass;
+    inertia_ = inertia;
+}
+
+void CentroidalPDController::run(
+    Eigen::Ref<const Eigen::Vector3d> kc,
+    Eigen::Ref<const Eigen::Vector3d> dc,
+    Eigen::Ref<const Eigen::Vector3d> kb,
+    Eigen::Ref<const Eigen::Vector3d> db,
+    Eigen::Ref<const Eigen::Vector3d> com,
+    Eigen::Ref<const Eigen::Vector3d> com_des,
+    Eigen::Ref<const Eigen::Vector3d> vcom,
+    Eigen::Ref<const Eigen::Vector3d> vcom_des,
+    Eigen::Ref<const Eigen::Vector4d> ori,
+    Eigen::Ref<const Eigen::Vector4d> ori_des,
+    Eigen::Ref<const Eigen::Vector3d> angvel,
+    Eigen::Ref<const Eigen::Vector3d> angvel_des
+) 
+{
+    /*************************************************************************/
+    // Compute the linear part of the wrench.
+
+    /*---------- computing position error ----*/
+    pos_error_.array() = com_des.array() - com.array();
+    vel_error_.array() = vcom_des.array() - vcom.array();
+    /*---------- computing tourques ----*/
+
+    wrench_.head<3>().array() =
+        mass_*(pos_error_.array()*kc.array() + vel_error_.array()*dc.array());
+
+    /*************************************************************************/
+    // Compute the angular part of the wrench.
+    des_ori_quat_.w() = ori_des[3];
+    des_ori_quat_.vec()[0] = ori_des[0];
+    des_ori_quat_.vec()[1] = ori_des[1];
+    des_ori_quat_.vec()[2] = ori_des[2];
+
+    ori_quat_.w() = ori[3];
+    ori_quat_.vec()[0] = ori[0];
+    ori_quat_.vec()[1] = ori[1];
+    ori_quat_.vec()[2] = ori[2];
+
+    des_ori_se3_ = des_ori_quat_.toRotationMatrix();
+    ori_se3_ = ori_quat_.toRotationMatrix();
+
+    ori_error_se3_ = des_ori_se3_.transpose() * ori_se3_;
+    ori_error_quat_ = ori_error_se3_;
+
+    //todo: multiply as matrix
+
+    ori_error_[0] = -2.0*((ori_error_quat_.w()*ori_error_quat_.vec()[0] * kb[0]) + (kb[2] - kb[1])*(ori_error_quat_.vec()[1]*ori_error_quat_.vec()[2]));
+    ori_error_[1] = -2.0*((ori_error_quat_.w()*ori_error_quat_.vec()[1] * kb[1]) + (kb[0] - kb[2])*(ori_error_quat_.vec()[0]*ori_error_quat_.vec()[2]));
+    ori_error_[2] = -2.0*((ori_error_quat_.w()*ori_error_quat_.vec()[2] * kb[2]) + (kb[1] - kb[0])*(ori_error_quat_.vec()[1]*ori_error_quat_.vec()[0]));
+
+    /*---------- computing ang error ----*/
+
+    wrench_.tail<3>().array() = 
+        db.array() * inertia_.array() * (angvel_des.array() - angvel.array()) + 
+        ori_error_.array();
+}
+
+Vector6d& CentroidalPDController::get_wrench()
+{
+    return wrench_;
+}
+
+}  // namespace blmc_controllers

--- a/srcpy/blmc_controllers.cpp
+++ b/srcpy/blmc_controllers.cpp
@@ -13,6 +13,7 @@ namespace blmc_controllers{
 
 void bind_impedance_controller(pybind11::module &module);
 void bind_centroidal_pd_controller(pybind11::module &module);
+void bind_centroidal_force_qp_controller(pybind11::module& module);
 
 PYBIND11_MODULE(py_blmc_controllers, m) {
   m.doc() = R"pbdoc(
@@ -27,6 +28,7 @@ PYBIND11_MODULE(py_blmc_controllers, m) {
 
   bind_impedance_controller(m);
   bind_centroidal_pd_controller(m);
+  bind_centroidal_force_qp_controller(m);
 }
 
 } // namespace blmc_controllers

--- a/srcpy/blmc_controllers.cpp
+++ b/srcpy/blmc_controllers.cpp
@@ -12,8 +12,9 @@
 namespace blmc_controllers{
 
 void bind_impedance_controller(pybind11::module &module);
+void bind_centroidal_pd_controller(pybind11::module &module);
 
-PYBIND11_MODULE(blmc_controllers, m) {
+PYBIND11_MODULE(py_blmc_controllers, m) {
   m.doc() = R"pbdoc(
         blmc_controllers python bindings
         ---------------------------------
@@ -25,6 +26,7 @@ PYBIND11_MODULE(blmc_controllers, m) {
     )pbdoc";
 
   bind_impedance_controller(m);
+  bind_centroidal_pd_controller(m);
 }
 
 } // namespace blmc_controllers

--- a/srcpy/centroidal_force_qp_controller.cpp
+++ b/srcpy/centroidal_force_qp_controller.cpp
@@ -1,0 +1,33 @@
+/**
+ * @file
+ * @license BSD 3-clause
+ * @copyright Copyright (c) 2020, New York University and Max Planck
+ * Gesellschaft
+ *
+ * @brief Python bindings for the CentroidalForceQPController class.
+ */
+
+#include "blmc_controllers/centroidal_force_qp_controller.hpp"
+#include <pybind11/pybind11.h>
+#include <pybind11/eigen.h>
+#include <pybind11/numpy.h>
+
+namespace py = pybind11;
+
+namespace blmc_controllers{
+
+void bind_centroidal_force_qp_controller(py::module& module)
+{
+    py::class_<CentroidalForceQPController>(module, "CentroidalForceQPController")
+        .def(py::init<>())
+
+        // Public methods.
+        .def("initialize",  &CentroidalForceQPController::initialize)
+        .def("run",  &CentroidalForceQPController::run)
+        .def("get_forces",
+             &CentroidalForceQPController::get_forces,
+             py::return_value_policy::reference)
+        ;
+}
+
+} // blmc_controllers

--- a/srcpy/centroidal_pd_controller.cpp
+++ b/srcpy/centroidal_pd_controller.cpp
@@ -1,0 +1,33 @@
+/**
+ * @file
+ * @license BSD 3-clause
+ * @copyright Copyright (c) 2020, New York University and Max Planck
+ * Gesellschaft
+ *
+ * @brief Python bindings for the CentroidalPDController class.
+ */
+
+#include "blmc_controllers/centroidal_pd_controller.hpp"
+#include <pybind11/pybind11.h>
+#include <pybind11/eigen.h>
+#include <pybind11/numpy.h>
+
+namespace py = pybind11;
+
+namespace blmc_controllers{
+
+void bind_centroidal_pd_controller(py::module& module)
+{
+    py::class_<CentroidalPDController>(module, "CentroidalPDController")
+        .def(py::init<>())
+
+        // Public methods.
+        .def("initialize",  &CentroidalPDController::initialize)
+        .def("run",  &CentroidalPDController::run)
+        .def("get_wrench",
+             &CentroidalPDController::get_wrench,
+             py::return_value_policy::reference)
+        ;
+}
+
+} // blmc_controllers

--- a/srcpy/impedance_controller.cpp
+++ b/srcpy/impedance_controller.cpp
@@ -11,7 +11,6 @@
 #include <pybind11/pybind11.h>
 #include <pybind11/eigen.h>
 #include <pybind11/numpy.h>
-#include <boost/python.hpp>
 
 namespace py = pybind11;
 
@@ -30,7 +29,7 @@ void bind_impedance_controller(py::module& module)
                 const std::string& end_frame_name)
              {
                  const pinocchio::Model& pinocchio_model =
-                    boost::python::extract<const pinocchio::Model&>(model.ptr());
+                    model.cast<const pinocchio::Model&>();
                  obj.initialize(pinocchio_model, root_frame_name, end_frame_name);
                  return ;
              })
@@ -45,14 +44,11 @@ void bind_impedance_controller(py::module& module)
                        py::object py_feed_forward_force)
                     {
                         const pinocchio::SE3& desired_end_frame_placement = 
-                            boost::python::extract<const pinocchio::SE3&>(
-                                py_desired_end_frame_placement.ptr());
+                            py_desired_end_frame_placement.cast<const pinocchio::SE3&>();
                         const pinocchio::Motion& desired_end_frame_velocity = 
-                            boost::python::extract<const pinocchio::Motion&>(
-                                py_desired_end_frame_velocity.ptr());
+                            py_desired_end_frame_velocity.cast<const pinocchio::Motion&>();
                         const pinocchio::Force& feed_forward_force = 
-                            boost::python::extract<const pinocchio::Force&>(
-                                py_feed_forward_force.ptr());
+                            py_feed_forward_force.cast<const pinocchio::Force&>();
                         obj.run(robot_configuration, robot_velocity,
                                 gain_proportional, gain_derivative,
                                 gain_feed_forward_force,

